src/test/java/org/mockito/internal/stubbing/defaultanswers/ReturnsSmartNullsTest.java

```diff
@@ -85,4 +90,36 @@ public void should_print_the_parameters_on_SmartNullPointerException_message() t
                 .hasMessageContaining("lumpa");
         }
     }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void generic_return_type_erasure() throws Throwable {
+        class S<T> {
+            public T f() {
+                return null;
+            }
+        }
+
+        S<Boolean> s = Mockito.mock(S.class, org.mockito.Answers.RETURNS_SMART_NULLS);
```

Comment:
This cannot be fixed without providing the generic argument to Mockito. We would need to add something like:
```java
mockGeneric(S.class, Boolean.class, org.mockito.Answers.RETURNS_SMART_NULLS)
```
Byte Buddy is capable of subclassing a generic type where anything down the chain would automatically adapt the actual type.

Currently, this is however a case of heap-pollution which the Java compiler warns against.